src/io/perun.clj

```diff
@@ -105,74 +145,103 @@
+(defn- diff-tree
+  [tree props]
+  (let [->map #(select-keys % props)]
+    (reduce-kv #(assoc %1 %2 (->map %3)) {} tree)))
+
+(defn- diff*
+  [{t1 :tree :as before} {t2 :tree :as after} props]
+  (if-not before
+    {:added   after
+     :removed (assoc after :tree {})
+     :changed (assoc after :tree {})}
+    (let [props        (or (seq props) [:id])
+          d1           (diff-tree t1 props)
+          d2           (diff-tree t2 props)
+          [x y z]      (map (comp set keys) (data/diff d1 d2))
+          changed-keys (set/union (set/intersection x y)
+                                  (set/intersection (set/union x y) z))]
+      {:added   (->> (set/difference y x changed-keys) (select-keys t2) (assoc after :tree))
+       :removed (->> (set/difference x y changed-keys) (select-keys t1) (assoc after :tree))
+       :changed (->> changed-keys                      (select-keys t2) (assoc after :tree))})))
+
+(defn- diff
+  [before after & props]
+  (let [{:keys [added changed]}
+        (diff* before after props)]
+    (update-in added [:tree] merge (:tree changed))))
 
 (defn content-pre-wrap
   "Wrapper for input parsing tasks. Calls `parse-form` on new or changed
-  files with extensions in `file-exts`, adds `tracer` to `:io.perun/trace`
-  and writes html files for subsequent tasks to process, if desired. Pass
+  files with extensions in `extensions`, adds `tracer` to `:io.perun/trace`
+  and writes files for subsequent tasks to process, if desired.
+  `output-extension` can be used to indicate the type of file the task produces,
+  or pass `nil` to overwrite with a file of the same name and extension. Pass
   `pod` if one is needed for parsing"
-  [parse-form file-exts tracer & [pod]]
-  (let [tmp     (boot/tmp-dir!)
-        prev-fs (atom nil)]
+  [parse-form-fn extensions output-extension tracer options & [pod]]
+  (let [tmp  (boot/tmp-dir!)
+        prev (atom {})]
     (boot/with-pre-wrap fileset
-      (let [changed-files (->> (boot/fileset-diff @prev-fs fileset :hash)
-                               boot/ls
-                               (boot/by-ext file-exts)
-                               add-filedata)
-            changed-meta (trace tracer
-                                (if pod
-                                  (pod/with-call-in @pod ~(parse-form changed-files))
-                                  (eval (parse-form changed-files))))
-            input-fs (-> (if @prev-fs
-                           (pm/set-meta fileset (meta-by-ext @prev-fs file-exts))
-                           fileset)
+      (let [global-meta (pm/get-global-meta fileset)
+            prev-fs (:fs @prev)
+            parse-form (parse-form-fn (meta-by-ext (boot/fileset-diff prev-fs fileset :hash) extensions))
+            ;; Change `diff` to boot/fileset-diff when
+            ;; https://github.com/boot-clj/boot/pull/566 is merged
+            changed-from-prev-meta (meta-by-ext (diff prev-fs fileset pm/+meta-key+) extensions)
```

Comment:
> That makes sense, though makes me wonder if the files where the Perun metadata has changed might need to be reparsed.

I've had some thoughts in this direction, actually.  I'll try to lay them out briefly.

1. `markdown`, and other content parsing tasks that are in progress or have been discussed, don't use Perun metadata in any meaningful way.
1. In a future PR (here: https://github.com/bhagany/perun/tree/markdown-asunder), I've split the `markdown` task into yaml and markdown-only subtasks, as you argued for here: https://github.com/hashobject/perun/pull/106
1. If that PR is merged, it's then possible to intelligently skip markdown parsing when only the yaml metadata has changed in a file.  I liked that, so I laid the groundwork here, which means diffing content and metadata separately.
1. However, I've also had thoughts about merging `content-pre-wrap` and `render-pre-wrap`, since they are so conceptually similar. If that happens, then skipping processing when metadata changes will no longer be feasible.
1. Adding a content parser that uses Perun metadata would also make the two diffs approach infeasible.

I guess I'm going with a do-the-most-efficient-thing-until-we-can't approach?